vortex-array/src/array/mod.rs

```diff
@@ -397,19 +393,33 @@ impl<V: VTable> Array for ArrayAdapter<V> {
             "Slice length mismatch {}",
             self.encoding_id()
         );
-        assert_eq!(
+        // Slightly more expensive, so only do this in debug builds.
+        debug_assert_eq!(
             sliced.dtype(),
             self.dtype(),
             "Slice dtype mismatch {}",
             self.encoding_id()
         );
 
-        if let Some(derived_stats) = derived_stats {
-            let mut stats = sliced.statistics().to_owned();
-            stats.combine_sets(&derived_stats, self.dtype())?;
-            for (stat, val) in stats.into_iter() {
-                sliced.statistics().set(stat, val)
-            }
+        // Propagate some stats from the original array to the sliced array.
+        if !sliced.is::<ConstantVTable>() {
+            self.statistics().with_iter(|iter| {
+                sliced.statistics().inherit(iter.filter(|stat| {
+                    match stat {
```

Comment:
nit: probably simpler to write `matches!(stat, Stat::IsConstant | Stat::IsSorted | Stat::IsStrictSorted) && value.as_bool().vortex_expect("must be a bool").unwrap_or_default()`